spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/DefaultLockRepository.java

```diff
@@ -174,17 +177,20 @@ public boolean acquire(String lock) {
 		}
 	}
 
+	@Transactional(propagation = Propagation.REQUIRES_NEW, readOnly = true)
```

Comment:
I wonder why do we need a dedicated transaction for `SELECT`?..

Reply:
Actually I didn't think about it. Just moved `@Transactional` method from the class to each method that executes some query. But now when you asked I researched it and concluded that new transaction is still required here even for methods that only read from DB.

If current transaction's isolation level is REPEATABLE_READ or SERIALIZABLE then after 1st `SELECT` all further `SELECT`s will return the same data, even if DB records were already modified/deleted by concurrent transactions.

This means that if first call to `this.client.isAcquired(path)`  returned `true` (if was called from user's transaction with REPEATABLE_READ isolation level), then all next calls will also return `true` while user's transaction is still active.

This means that `registry.expireUnusedOlderThan(..)` will never expire anything as long as user's transaction is still active.

I written a small test that confirms this.

```java
	@Transactional(isolation = Isolation.REPEATABLE_READ)
	@Test
	public void testIsAcquiredFromRepeatableReadTransaction() {
		// Make sure a transaction with REPEATABLE_READ isolation level is active
		assertThat(TransactionSynchronizationManager.isActualTransactionActive()).isTrue();
		assertThat(TransactionSynchronizationManager.getCurrentTransactionIsolationLevel())
				.isEqualTo(Connection.TRANSACTION_REPEATABLE_READ);

		Lock lock = this.registry.obtain("foo");
		String path = TestUtils.getPropertyValue(lock, "path", String.class);
		lock.lock();
		try {
			assertThat(this.client.isAcquired(path)).isTrue();
		}
		finally {
			lock.unlock();
		}
		assertThat(this.client.isAcquired(path)).isFalse();
	}
```

This test fails when this `SELECT` doesn't execute in a new transaction.

I also tried to use

```java
this.registry.expireUnusedOlderThan(-1000);
assertThat(TestUtils.getPropertyValue(this.registry, "locks", Map.class).size()).isEqualTo(1); // and then isEqualTo(0) after unlock()
```

instead of

```
assertThat(this.client.isAcquired(path)).isTrue(); // and then isFalse()
```

and it also fails.

Should I add this test too? Maybe I should replace `lock.lock()`/`lock.unlock()` with direct usage of `DefaultLockRepository` methods. This case is actually already covered by test that checks transaction's suspension count. This additional test just makes the intention more clear for the `SELECT` case.

Reply:
Actually I thought about it more and maybe it's worth to set `Propagation.NOT_SUPPORTED` on `boolean isAcquired(String lock)`? This way the test still passes as the problem is not the absence of a new transaction but presence of existing transaction. Executing SELECT without opening and closing a new transaction will make this method faster in theory. It will still be executed in a new DB connection and current transaction will still be suspended, though.

Reply:
Good. Thanks for the report. Didn't think about the cause behind `REPEATABLE_READ`.
Let's have that TX on the SELECT and let's include such a test into our suite!

Reply:
Yeah... Let's leave it transactional for consistency and clear separation of concerns that the state around lock record in DB has to be volatile in terms of Java 😸

Reply:
Yeah, it's pretty subtle. The problem when user's transaction is `SERIALIZABLE` will be even worse, as `SELECT` will acquire shared lock on DB record and all concurrent transactions that trying to do something with this DB record will be blocked waiting for this shared lock to be released (bye-bye `tryLock()`). And it will be released only after user's transaction will be committed. 🙈 

Btw, I removed the `JdbcLockRegistry` test and added dedicated tests for `DefaultLockRepository`.